src/plan_agg_bookend.c

```diff
@@ -0,0 +1,748 @@
+ * node. Param node passes output value from the subquery.
+ *
+ */
+void
+replace_aggref_in_tlist(MinMaxAggPath *minmaxagg_path)
+{
+	MutatorContext context;
+
+	context.mm_path = minmaxagg_path;
+
+	((Path *) minmaxagg_path)->pathtarget->exprs = (List *) mutate_aggref_node((Node *) ((Path *) minmaxagg_path)->pathtarget->exprs, (void *) &context);
+}
+
+/* Stores map of type oid -> sortop oids for a particular function */
+typedef struct FuncSortOp
+{
+	Oid			func_oid;
+	HTAB	   *sort_op_map;
+} FuncSortOp;
+
+typedef struct SortOpInfo
+{
+	Oid			sort_oid;		/* map key */
+
+	Oid			sortop_oid;
+} SortOpInfo;
+
+HASHCTL		sortop_map_hashctl = {
+	.keysize = sizeof(Oid),
+	.entrysize = sizeof(SortOpInfo)
+};
+
+static Oid	first_last_arg_types[] = {ANYELEMENTOID, ANYOID};
+
+struct FuncSortOp first_sort_op_info = {.func_oid = InvalidOid};
+struct FuncSortOp last_sort_op_info = {.func_oid = InvalidOid};
+
+static Oid
+get_function_oid(char *name, int nargs, Oid arg_types[])
+{
+	FuncCandidateList func_candidates;
+
+	func_candidates = FuncnameGetCandidates(list_make1(makeString(name)), nargs, NIL, false, false, false);
+	while (func_candidates != NULL)
+	{
+		if (func_candidates->nargs == nargs && function_types_equal(func_candidates->args, arg_types, nargs))
+			return func_candidates->oid;
+		func_candidates = func_candidates->next;
+	}
+	elog(ERROR, "failed to find function %s with %d args", name, nargs);
+}
+
+static FuncSortOp *
+initialize_func_sortop(FuncSortOp *func_sortop, char *name, int nargs, Oid arg_types[])
+{
+	func_sortop->func_oid = get_function_oid(name, nargs, arg_types);
+	func_sortop->sort_op_map = hash_create(name, 10, &sortop_map_hashctl, HASH_ELEM | HASH_BLOBS);
+	return func_sortop;
+}
+
+static FuncSortOp *
+get_func_sort_op(Oid func_oid)
+{
+	if (first_sort_op_info.func_oid == InvalidOid)
+		initialize_func_sortop(&first_sort_op_info, "first", 2, first_last_arg_types);
+	if (last_sort_op_info.func_oid == InvalidOid)
+		initialize_func_sortop(&last_sort_op_info, "last", 2, first_last_arg_types);
+	if (first_sort_op_info.func_oid == func_oid)
+		return &first_sort_op_info;
+	if (last_sort_op_info.func_oid == func_oid)
+		return &last_sort_op_info;
+	return NULL;
+}
+
+static Oid
+get_sort_op_oid(Oid func_oid, Oid sort_oid)
+{
+	FuncSortOp *func_sort_op = get_func_sort_op(func_oid);
+
+	if (func_sort_op == NULL)
+		return InvalidOid;
+	bool		existing = false;
+	SortOpInfo *sort_op_info;
+
+	if (func_sort_op->sort_op_map == NULL)
+	{
+		func_sort_op->sort_op_map = hash_create("sort_op_map", 10, &sortop_map_hashctl, HASH_ELEM | HASH_BLOBS);
+	}
+	sort_op_info = (SortOpInfo *) hash_search(func_sort_op->sort_op_map, &sort_oid, HASH_ENTER, &existing);
+	if (existing)
+		return sort_op_info->sortop_oid;
+	else
+	{
+		sort_op_info->sort_oid = sort_oid;
+		if (func_sort_op == &first_sort_op_info)
+			sort_op_info->sortop_oid = OpernameGetOprid(list_make1(makeString("<")), sort_oid, sort_oid);
```

Comment:
why are we not using `get_opfamily_member` as I mentioned last time?

Reply:
if we use `tce` and `get_opfamily_member` those are both cached lookups. Therefore, i'd rather not do our own cache via a hashtable here. I don't see the point.

Reply:
Sure. I brought up that point about our own cache vs PostgreSql type caching during our sync. 
I am up for avoiding adding more complexity if we don't have to, but I am also not completely sure what guarantees we get from PostgreSQL type caching (eg. cache expiration)